### PR TITLE
Clean composer promises

### LINKED ISSUE #1438: Discussion composer doesn't open after login, promise is useless
I noticed that in `IndexPage.newDiscussion()` a nice system is used to allow opening the login modal and returning the composer once it's ready.

https://github.com/flarum/core/blob/4df0101f5644312773c40638ea3d54ac86df462b/js/forum/src/components/IndexPage.js#L344

Except it doesn't (can't) work, because the login modal refreshes the page once logged in.

I don't suppose it's planned to remove that hard-refresh as a lot of stuff might need to be reloaded via another way.

Given that, the code to handle the login (`onlogin`, which by the way does not appear to have ever existed !?) could be removed to clean the code.

I also think the use of a promise is completely useless. There could simply be a method returning the new composer that can be extended via `extend()` like most of Flarum methods.

I don't think there's any way to still open the composer after login, given the full refresh, so the `newDiscussion` method would simply be a switch between open composer or open login without any callback.

I was going to write a PR for that but the change would have effect on other extensions using that method/promise.

So far the only extension I know is using that is the tags extension. It could be easily replaced with an `extend()` case as described above.

Is there anything else I should take into consideration to make that PR ? Does that promise serve another purpose I've not discovered ?

### PR DESCRIPTION
**Pull request by clarkwinkelmann**
Fixes #1438 

After deeper investigation I noticed the promise from `utils/PostControls.editAction()` is actually used in the mentions extension (https://github.com/flarum/flarum-ext-mentions/blob/6f0917cff69f662bde3e454ea1f998e4184e8320/js/forum/src/utils/reply.js#L32) and it would be a bit complicated or way less pretty to replace that with a solution that doesn't involve promises.

So I think instead of my previous suggestion to drop all promises, it's better to instead make the usage of promises consistent across all composer actions.

This PR does the following:

- Removed unused code that was supposed to open the login modal and return to the composer
- Made all "composer action" methods return a promise with receive the composer component as parameter
- Always resolve or reject the promise. Not doing so might keep Mithril waiting for an onclick event to finish. Doesn't seem to be an issue, but I guess it's better to always finish a promise properly
- Renamed the method to open the new discussion modal so it's more consistent with the other composer method names

Another PR makes the changes to the tags extension to work with the new method https://github.com/flarum/flarum-ext-tags/pull/51

### CODE REVIEW

**Comment by franzliedke** [comment]
@clarkwinkelmann Could you rebase and fix the conflicts, please? :pray:

**Comment by clarkwinkelmann** [comment]
@franzliedke I rebased the PR, hope I didn't break anything. I've not tested the code again, I just checked the rebase put the code at the right place.

@tobscure I understand your reasoning. I don't really know what the best option is though :man_shrugging:

**Comment by franzliedke** [comment]
> Always resolve or reject the promise. Not doing so might keep Mithril waiting for an onclick event to finish.

Doesn't this just mean that we would leak a bit of memory?

File: js/src/forum/utils/PostControls.js [hunk 1]
```diff
@@ -111,10 +111,20 @@ export default {
...
+    const deferred = m.deferred();
+
+    const component = new EditPostComposer({ post: this });
+
+    app.composer.load(component);
     app.composer.show();
+
+    deferred.resolve(component);
+
+    return deferred.promise;
```
[the remark this answers is not in the document]
Comment (tobscure) [reply js/src/forum/utils/PostControls.js]: Yeah, unfortunately Mithril 0.2's promises have a proprietary interface https://mithril.js.org/archive/v0.2.5/mithril.deferred.html

**Pull request merged**